File: app/src/main/java/com/mxt/anitrend/util/IntentBundleUtil.java
package com.mxt.anitrend.util;

import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.ShareCompat;
import android.text.TextUtils;

import java.util.regex.Matcher;

/**
 * Created by max on 2017/12/01.
 * Intent data formatter and helper methods
 */

public class IntentBundleUtil {

    private ShareCompat.IntentReader sharedIntent;
    private Matcher deepLinkMatcher;
    private String INTENT_ACTION;

    public IntentBundleUtil(Intent intent, FragmentActivity context) {
        if(intent != null) {
            if(intent.hasExtra(KeyUtil.arg_shortcut_used) && Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1)
                ShortcutUtil.reportShortcutUsage(context, intent.getIntExtra(KeyUtil.arg_shortcut_used, KeyUtil.SHORTCUT_SEARCH));
            checkIntentData(intent, context);
        }
    }

    private boolean hasAction(Intent intent) {
        return (INTENT_ACTION = intent.getAction()) != null;
    }

    private @Nullable Uri hasData(Intent intent) {
        return intent.getData();
    }

    private @Nullable String[] hasDepth(String key) {
        if(key.contains("/"))
            return key.split("/");
        return null;
    }

    private void injectIntentParams(Intent intent) {
        String type = deepLinkMatcher.group(1);
        int groupLimit = deepLinkMatcher.groupCount();
        String lastKey = deepLinkMatcher.group(groupLimit);
        String[] splitKeys;
        switch (type) {
            case KeyUtil.DEEP_LINK_USER:
                if(TextUtils.isDigitsOnly(lastKey))
                    intent.putExtra(KeyUtil.arg_id, Long.valueOf(lastKey));
                else {
                    if (lastKey.contains("/"))
                        lastKey = lastKey.replace("/", "");
                    intent.putExtra(KeyUtil.arg_userName, lastKey);
                }
                break;
            case KeyUtil.DEEP_LINK_MANGA:
                if ((splitKeys = hasDepth(lastKey)) != null)
                    intent.putExtra(KeyUtil.arg_id, Long.valueOf(splitKeys[0]));
                else
                    intent.putExtra(KeyUtil.arg_id, Long.valueOf(lastKey));
                intent.putExtra(KeyUtil.arg_mediaType, KeyUtil.MANGA);
                    break;
            case KeyUtil.DEEP_LINK_ANIME:
                if ((splitKeys = hasDepth(lastKey)) != null)
                    intent.putExtra(KeyUtil.arg_id, Long.valueOf(splitKeys[0]));
                else
                    intent.putExtra(KeyUtil.arg_id, Long.valueOf(lastKey));
                intent.putExtra(KeyUtil.arg_mediaType, KeyUtil.ANIME);
                    break;
            case KeyUtil.DEEP_LINK_CHARACTER:
                if ((splitKeys = hasDepth(lastKey)) != null)
                    intent.putExtra(KeyUtil.arg_id, Long.valueOf(splitKeys[0]));
                else
                    intent.putExtra(KeyUtil.arg_id, Long.valueOf(lastKey));
                break;
            case KeyUtil.DEEP_LINK_ACTOR:
                if ((splitKeys = hasDepth(lastKey)) != null)
                    intent.putExtra(KeyUtil.arg_id, Long.valueOf(splitKeys[0]));
                else
                    intent.putExtra(KeyUtil.arg_id, Long.valueOf(lastKey));
                break;
            case KeyUtil.DEEP_LINK_STAFF:
                if ((splitKeys = hasDepth(lastKey)) != null)
                    intent.putExtra(KeyUtil.arg_id, Long.valueOf(splitKeys[0]));
                else
                    intent.putExtra(KeyUtil.arg_id, Long.valueOf(lastKey));
                break;
            case KeyUtil.DEEP_LINK_STUDIO:
                if ((splitKeys = hasDepth(lastKey)) != null)
                    intent.putExtra(KeyUtil.arg_id, Long.valueOf(splitKeys[0]));
                else
                    intent.putExtra(KeyUtil.arg_id, Long.valueOf(lastKey));
                break;
        }
    }

    private void checkIntentData(Intent intent, FragmentActivity context) {
        Uri INTENT_DATA;
        if (hasAction(intent) && INTENT_ACTION.equals(Intent.ACTION_SEND))
            sharedIntent = ShareCompat.IntentReader.from(context);
        else
            if ((INTENT_DATA = hasData(intent)) != null)
                if((deepLinkMatcher = RegexUtil.INSTANCE.findIntentKeys(INTENT_DATA.getPath())) != null)
                    injectIntentParams(intent);
    }

    public @Nullable ShareCompat.IntentReader getSharedIntent() {
        return sharedIntent;
    }
}

File: app/src/main/java/com/mxt/anitrend/util/IntentBundleUtil.kt
package com.mxt.anitrend.util

import android.content.Intent
import android.net.Uri
import android.os.Build
import androidx.annotation.VisibleForTesting
import androidx.fragment.app.FragmentActivity
import androidx.core.app.ShareCompat
import android.text.TextUtils

import java.util.regex.Matcher

/**
 * Created by max on 2017/12/01.
 * Intent data formatter and helper methods
 */

class IntentBundleUtil(private val intent: Intent) {

    var sharedIntent: ShareCompat.IntentReader? = null

    private val deepLinkMatcher: Matcher? by lazy {
        RegexUtil.findIntentKeys(intentData?.path)
    }

    private val intentAction: String? = intent.action

    private val intentData: Uri? = intent.data

    private fun hasDepth(key: String): Array<String>? {
        return if (key.contains("/")) 
            key.split("/".toRegex())
                    .dropLastWhile {
                        it.isEmpty()
                    }.toTypedArray()
        else null
    }

    private fun injectIntentParams() {
        deepLinkMatcher?.also {
            val type = it.group(1)
            val groupLimit = it.groupCount()

            var lastKey = it.group(groupLimit)
            val splitKeys: Array<String>? = hasDepth(lastKey)

            when (type) {
                KeyUtil.DEEP_LINK_ACTIVITY -> {
                    if (splitKeys != null)
                        intent.putExtra(KeyUtil.arg_id, splitKeys[0].toLong())
                    else
                        intent.putExtra(KeyUtil.arg_id, lastKey.toLong())
                }
                KeyUtil.DEEP_LINK_USER -> when {
                    TextUtils.isDigitsOnly(lastKey) -> intent.putExtra(KeyUtil.arg_id, lastKey.toLong())
                    else -> {
                        if (lastKey.contains("/"))
                            lastKey = lastKey.replace("/", "")
                        intent.putExtra(KeyUtil.arg_userName, lastKey)
                    }
                }
                KeyUtil.DEEP_LINK_MANGA -> {
                    if (splitKeys != null)
                        intent.putExtra(KeyUtil.arg_id, splitKeys[0].toLong())
                    else
                        intent.putExtra(KeyUtil.arg_id, lastKey.toLong())
                    intent.putExtra(KeyUtil.arg_mediaType, KeyUtil.MANGA)
                }
                KeyUtil.DEEP_LINK_ANIME -> {
                    if (splitKeys != null)
                        intent.putExtra(KeyUtil.arg_id, splitKeys[0].toLong())
                    else
                        intent.putExtra(KeyUtil.arg_id, lastKey.toLong())
                    intent.putExtra(KeyUtil.arg_mediaType, KeyUtil.ANIME)
                }
                KeyUtil.DEEP_LINK_CHARACTER -> if (splitKeys != null)
                    intent.putExtra(KeyUtil.arg_id, splitKeys[0].toLong())
                else
                    intent.putExtra(KeyUtil.arg_id, lastKey.toLong())

                KeyUtil.DEEP_LINK_ACTOR -> if (splitKeys != null)
                    intent.putExtra(KeyUtil.arg_id, splitKeys[0].toLong())
                else
                    intent.putExtra(KeyUtil.arg_id, lastKey.toLong())

                KeyUtil.DEEP_LINK_STAFF -> if (splitKeys != null)
                    intent.putExtra(KeyUtil.arg_id, splitKeys[0].toLong())
                else
                    intent.putExtra(KeyUtil.arg_id, lastKey.toLong())

                KeyUtil.DEEP_LINK_STUDIO -> if (splitKeys != null)
                    intent.putExtra(KeyUtil.arg_id, splitKeys[0].toLong())
                else
                    intent.putExtra(KeyUtil.arg_id, lastKey.toLong())
            }
        }
    }

    fun checkIntentData(context: FragmentActivity) {
        if (context.intent?.hasExtra(KeyUtil.arg_shortcut_used) == true && Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1)
            ShortcutUtil.reportShortcutUsage(context, context.intent.getIntExtra(KeyUtil.arg_shortcut_used, KeyUtil.SHORTCUT_SEARCH))

        if (!intentAction.isNullOrEmpty() && intentAction == Intent.ACTION_SEND)
            sharedIntent = ShareCompat.IntentReader.from(context)
        else if (deepLinkMatcher != null)
            injectIntentParams()
    }
}

